packages/flutter/lib/src/painting/borders.dart

```diff
@@ -259,6 +259,318 @@ class BorderSide {
+  /// the border will not need the text direction to paint itself.
+  ///
+  /// See also:
+  ///
+  ///  * [getInnerPath], which creates the path for the inner edge.
+  ///  * [Path.contains], which can tell if an [Offset] is within a [Path].
+  Path getOuterPath(Rect rect, { TextDirection textDirection });
+
+  /// Create a [Path] that describes the inner edge of the border.
+  ///
+  /// This path must not cross the path given by [getOuterPath] for the same
+  /// [Rect].
+  ///
+  /// To obtain a [Path] that describes the area of the border itself, set the
+  /// [Path.fillType] of the returned object to [PathFillType.evenOdd], and add
+  /// to this object the path returned from [getOuterPath] (using
+  /// [Path.addPath]).
+  ///
+  /// The `textDirection` argument must be provided and non-null if the border
+  /// has a text direction dependency (for example if it is expressed in terms
+  /// of "start" and "end" instead of "left" and "right"). It may be null if
+  /// the border will not need the text direction to paint itself.
+  ///
+  /// See also:
+  ///
+  ///  * [getOuterPath], which creates the path for the outer edge.
+  ///  * [Path.contains], which can tell if an [Offset] is within a [Path].
+  Path getInnerPath(Rect rect, { TextDirection textDirection });
+
+  /// Paints the border within the given [Rect] on the given [Canvas].
+  ///
+  /// The `textDirection` argument must be provided and non-null if the border
+  /// has a text direction dependency (for example if it is expressed in terms
+  /// of "start" and "end" instead of "left" and "right"). It may be null if
+  /// the border will not need the text direction to paint itself.
+  void paint(Canvas canvas, Rect rect, { TextDirection textDirection });
+
+  @override
+  String toString() {
+    return '$runtimeType()';
+  }
+}
+
+/// Represents the addition of two otherwise-incompatible borders.
+///
+/// The borders are listed from the outside to the inside.
+class _CompoundBorder extends ShapeBorder {
+  _CompoundBorder(this.borders) {
+    assert(borders != null);
+    assert(borders.length >= 2);
+    assert(!borders.any((ShapeBorder border) => border is _CompoundBorder));
+  }
+
+  final List<ShapeBorder> borders;
+
+  @override
+  EdgeInsetsGeometry get dimensions {
+    return borders.fold<EdgeInsetsGeometry>(
+      EdgeInsets.zero,
+      (EdgeInsetsGeometry previousValue, ShapeBorder border) {
+        return previousValue.add(border.dimensions);
+      },
+    );
+  }
+
+  @override
+  ShapeBorder add(ShapeBorder other, { bool reversed: false }) {
+    // This wraps the list of borders with "other", or, if "reversed" is true,
+    // wraps "other" with the list of borders.
+    // If "reversed" is false, "other" should end up being at the start of the
+    // list, otherwise, if "reversed" is true, it should end up at the end.
+    // First see if we can merge the new adjacent borders.
```

Comment:
super-ultra-pedanto-nit: First,

I'll be honest, I have no idea why... and admittedly it reads alright without it.

Reply:
fixed